File: sdk/elixir/test/support/loopback.ex
defmodule Matrix.Loopback do
  @moduledoc "In-process loopback fake host for the SDK tests (stdlib only)."

  @max_frame 1_048_576

  def send_frame(sock, map) do
    raw = :erlang.iolist_to_binary(:json.encode(map))
    :gen_tcp.send(sock, [<<byte_size(raw)::32-big>>, raw])
  end

  def read_frame(sock, timeout \\ 10_000) do
    with {:ok, <<n::32-big>>} <- recv_exact(sock, 4, timeout),
         {:ok, payload} <- recv_exact(sock, n, timeout) do
      :json.decode(payload)
    end
  end

  defp recv_exact(sock, n, timeout, acc \\ <<>>) do
    need = n - byte_size(acc)

    case :gen_tcp.recv(sock, need, timeout) do
      {:ok, data} ->
        acc = acc <> data
        if byte_size(acc) == n, do: {:ok, acc}, else: recv_exact(sock, n, timeout, acc)

      {:error, reason} ->
        {:error, reason}
    end
  end

  def env(type, body, rid \\ "r1") do
    %{
      "protocol" => "matrix.component",
      "version" => "0.1",
      "type" => type,
      "message_id" => "m1",
      "session_id" => "s1",
      "instance_id" => "1",
      "generation" => "1",
      "request_id" => rid,
      "body" => body
    }
  end

  def call_open(ticket, input) do
    %{
      "protocol" => "matrix.component",
      "version" => "0.1",
      "type" => "call.open",
      "message_id" => "m-#{ticket}",
      "session_id" => "s1",
      "instance_id" => "1",
      "generation" => "1",
      "request_id" => "r-#{ticket}",
      "body" => %{"ticket" => ticket, "capability" => "c@1", "input" => input}
    }
  end

  def dispose do
    env("lifecycle.dispose", %{"operation_id" => "op", "deadline_ms" => 100}, "rd1")
  end

  @doc """
  Serves one connection through the handshake, then runs `script`. Returns
  {sock_path, task} where task completes with the script.
  """
  def serve(features, bindings, script) do
    dir = Path.join(System.tmp_dir!(), "exunits-#{random()}")
    File.mkdir_p!(dir)
    path = Path.join(dir, "t.sock")
    {:ok, srv} =
      :gen_tcp.listen(0, [:binary, {:packet, :raw}, {:active, false}, {:ifaddr, {:local, String.to_charlist(path)}}])

    task =
      Task.async(fn ->
        {:ok, conn} = :gen_tcp.accept(srv)
        hello = read_frame(conn)
        true = hello["type"] == "hello"

        send_frame(conn, %{
          "protocol" => "matrix.component",
          "version" => "0.1",
          "type" => "welcome",
          "message_id" => "h1",
          "session_id" => "s1",
          "body" => %{"version" => "0.1", "max_frame" => @max_frame, "limits" => %{}, "features" => features}
        })

        reg = read_frame(conn)
        true = reg["type"] == "component.register"

        send_frame(conn, %{
          "protocol" => "matrix.component",
          "version" => "0.1",
          "type" => "registered",
          "message_id" => "r",
          "session_id" => "s1",
          "instance_id" => "1",
          "generation" => "1",
          "body" => %{"logical" => "t"}
        })

        act =
          env("lifecycle.activate", %{
            "operation_id" => "op",
            "manifest" => %{},
            "bindings" => [],
            "dependency_bindings" => bindings
          }, "q")

        send_frame(conn, act)
        lc = read_frame(conn)
        true = lc["type"] == "lifecycle.result"
        script.(conn)
        :gen_tcp.close(conn)
        :gen_tcp.close(srv)
        File.rm_rf(dir)
        :ok
      end)

    {path, task}
  end

  defp random, do: :crypto.strong_rand_bytes(4) |> Base.encode16(case: :lower)
end

File: sdk/csharp/test/SelfTest/Loopback.cs
// Loopback fake host for the C# SDK self-test (no packages).
using System.Buffers.Binary;
using System.Net.Sockets;
using System.Text;
using System.Text.Json;

static class Loopback
{
    public const int MaxFrame = 1024 * 1024;

    public static void Send(Socket conn, Dictionary<string, object?> msg)
    {
        var raw = JsonSerializer.SerializeToUtf8Bytes(msg);
        var frame = new byte[4 + raw.Length];
        BinaryPrimitives.WriteUInt32BigEndian(frame, (uint)raw.Length);
        Buffer.BlockCopy(raw, 0, frame, 4, raw.Length);
        conn.Send(frame);
    }

    public static JsonElement Read(Socket conn, int timeoutMs = 10000)
    {
        conn.ReceiveTimeout = timeoutMs;
        var hdr = ReceiveExact(conn, 4);
        uint n = BinaryPrimitives.ReadUInt32BigEndian(hdr);
        if (n == 0 || n > MaxFrame)
            throw new Exception($"bad length {n}");
        var payload = ReceiveExact(conn, (int)n);
        return JsonDocument.Parse(payload).RootElement.Clone();
    }

    private static byte[] ReceiveExact(Socket conn, int count)
    {
        var buf = new byte[count];
        int got = 0;
        while (got < count)
        {
            int n = conn.Receive(new ArraySegment<byte>(buf, got, count - got));
            if (n == 0)
                throw new Exception("eof");
            got += n;
        }
        return buf;
    }

    public static Dictionary<string, object?> Env(string type, Dictionary<string, object?> body)
    {
        return new Dictionary<string, object?>
        {
            ["protocol"] = "matrix.component",
            ["version"] = "0.1",
            ["type"] = type,
            ["message_id"] = "m1",
            ["session_id"] = "s1",
            ["instance_id"] = "1",
            ["generation"] = "1",
            ["request_id"] = "r1",
            ["body"] = body,
        };
    }

    public static Dictionary<string, object?> CallOpen(string ticket, object? input)
    {
        return new Dictionary<string, object?>
        {
            ["protocol"] = "matrix.component",
            ["version"] = "0.1",
            ["type"] = "call.open",
            ["message_id"] = "m-" + ticket,
            ["session_id"] = "s1",
            ["instance_id"] = "1",
            ["generation"] = "1",
            ["request_id"] = "r-" + ticket,
            ["body"] = new Dictionary<string, object?>
            {
                ["ticket"] = ticket,
                ["capability"] = "c@1",
                ["input"] = input ?? new Dictionary<string, object?>(),
            },
        };
    }

    public static Dictionary<string, object?> DisposeFrame()
    {
        return new Dictionary<string, object?>
        {
            ["protocol"] = "matrix.component",
            ["version"] = "0.1",
            ["type"] = "lifecycle.dispose",
            ["message_id"] = "d1",
            ["session_id"] = "s1",
            ["instance_id"] = "1",
            ["generation"] = "1",
            ["request_id"] = "rd1",
            ["body"] = new Dictionary<string, object?> { ["operation_id"] = "op", ["deadline_ms"] = 100 },
        };
    }

    /// <summary>
    /// Serves one connection through the handshake, then runs script.
    /// Returns the socket path and a task completing with the script.
    /// </summary>
    public static (string, Task) Serve(string[] features, object[] bindings, Action<Socket> script)
    {
        var dir = Directory.CreateDirectory(
            Path.Combine(Path.GetTempPath(), "csunits-" + Path.GetRandomFileName())).FullName;
        var sockPath = Path.Combine(dir, "t.sock");
        var srv = new Socket(AddressFamily.Unix, SocketType.Stream, ProtocolType.Unspecified);
        srv.Bind(new UnixDomainSocketEndPoint(sockPath));
        srv.Listen(1);
        var done = Task.Run(() =>
        {
            try
            {
                using var conn = srv.Accept();
                var hello = Read(conn);
                if (hello.GetProperty("type").GetString() != "hello")
                    throw new Exception("expected hello");
                Send(conn, new Dictionary<string, object?>
                {
                    ["protocol"] = "matrix.component",
                    ["version"] = "0.1",
                    ["type"] = "welcome",
                    ["message_id"] = "h1",
                    ["session_id"] = "s1",
                    ["body"] = new Dictionary<string, object?>
                    {
                        ["version"] = "0.1",
                        ["max_frame"] = MaxFrame,
                        ["limits"] = new Dictionary<string, object?>(),
                        ["features"] = features,
                    },
                });
                var reg = Read(conn);
                if (reg.GetProperty("type").GetString() != "component.register")
                    throw new Exception("expected register");
                Send(conn, new Dictionary<string, object?>
                {
                    ["protocol"] = "matrix.component",
                    ["version"] = "0.1",
                    ["type"] = "registered",
                    ["message_id"] = "r",
                    ["session_id"] = "s1",
                    ["instance_id"] = "1",
                    ["generation"] = "1",
                    ["body"] = new Dictionary<string, object?> { ["logical"] = "t" },
                });
                var act = Env("lifecycle.activate", new Dictionary<string, object?>
                {
                    ["operation_id"] = "op",
                    ["manifest"] = new Dictionary<string, object?>(),
                    ["bindings"] = Array.Empty<object>(),
                    ["dependency_bindings"] = bindings,
                });
                act["request_id"] = "q";
                Send(conn, act);
                var lc = Read(conn);
                if (lc.GetProperty("type").GetString() != "lifecycle.result")
                    throw new Exception("expected lifecycle.result");
                script(conn);
            }
            finally
            {
                srv.Close();
                try { Directory.Delete(dir, recursive: true); } catch { /* best effort */ }
            }
        });
        return (sockPath, done);
    }
}
